### PR TITLE
Added new lint "path_join_correction"

### PR DESCRIPTION
**Pull request by ofeeg**
Thank you for making Clippy better!

We're collecting our changelog from pull request descriptions.
If your PR only includes internal changes, you can just write
`changelog: none`. Otherwise, please write a short comment
explaining your change.

It's also helpful for us that the lint name is put within backticks (`` ` ` ``),
and then encapsulated by square brackets (`[]`), for example:
```
changelog: [`lint_name`]: your change
```

If your PR fixes an issue, you can add `fixes #issue_number` into this
PR description. This way the issue will be automatically closed when
your PR is merged.

If you added a new lint, here's a checklist for things that will be
checked during review or continuous integration.

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.

Delete this line and everything above before opening your PR.

---

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: Added lint resolving issue #10655

### CODE REVIEW

**Review by xFrednet — COMMENTED** [review]
Welcome to Clippy :wave:. This taught me something new about the behavior of `Path::join()`. I think this is an excellent lint, which should be warn-by-default. You already have a good start and added the lint at the right location. I've left comments, how to improve the implementation.

If you require any further assistance, you're very welcome to reach out :)

**Comment by xFrednet** [comment]
> Hi! Sorry I have not looked at my computer for a couple of days and the app I uses on my phone to interact with github did not notify me of any of these replies. I've read them all now and will start on implementing each of your requests one at a time.

No problem at all, my studies are currently quite time-consuming. Having some time between reviews helps me as well. Just let me know when it's ready for the next review or when you need some assistance!

> EDIT: Also, thank you for letting me know about the email address clashing. I use magit in emacs and forgot to set a valid email address.

No problem! I also mention it, in case, the email/name is supposed to be private. I once commited something under my real name, before I had it linked to my GH and luckily caught it before it landed on master :)

**Comment by xFrednet** [comment]
Oh, I totally missed the PR update. Sorry, I'll give you a review soon.

@blyxyas Do you maybe want to give this one a go? :)

File: clippy_lints/src/methods/path_join_correction.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use rustc_lint::LateContext;
+use rustc_span::{symbol::sym::Path, Span};
+
+use super::PATH_JOIN_CORRECTION;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, join_arg: &'tcx Expr<'tcx>, span: Span) {
+    let ty = cx.typeck_results().expr_ty(expr);
+    if_chain!(
+     if is_type_diagnostic_item(cx, ty, Path);
+     let applicability = Applicability::MachineApplicable;
```
Comment (blyxyas) [inline clippy_lints/src/methods/path_join_correction.rs:14]: No need to bind the applicability to a variable, don't worry :)
(Personal taste here, it's still fine)

File: tests/ui/path_join_correction.rs
Comment (blyxyas) [inline tests/ui/path_join_correction.rs:1]: Could you also commit the `.stderr` version and `.fixed` of this files? You can get those by using `cargo dev bless` after a `cargo test <blah blah>`.

**Review by blyxyas — COMMENTED** [review]
Very good for a first contribution! Thanks for joining the project! :heart:

File: tests/ui/join_absolute_path.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// run-rustfix
```
Comment (blyxyas) [inline tests/ui/join_absolute_path.rs:1]:
```suggestion
//@run-rustfix
```

**Review by blyxyas — CHANGES_REQUESTED** [review]
Headers now use `//@`

**Comment by blyxyas** [comment]
It seems like blessing isn't working... Could you describe your workflow?

**Comment by ofeeg** [comment]
> It seems like blessing isn't working... Could you describe your workflow?

Not sure. Eventually I noticed that .fixed and .stderr files weren't being created. I use emacs and with this I just did a grep'd the old name and replaced all instances of the old name with a new name using sed.  I wonder if I should just refork clippy and copy all the work I've done to the new fork.

**Comment by blyxyas** [comment]
Just a `cargo test` and then a `cargo dev bless` should fix it :thinking:

**Comment by ofeeg** [comment]
> Just a `cargo test` and then a `cargo dev bless` should fix it thinking

I did that and those files just don't get created. Also, this still fails. Unsure what possibly happened, but it's been a while since I last updated this and for all I can remember most of what I did was just keep doing cargo commands until the thing worked. I probably bugged it out at some point.

**Comment by blyxyas** [comment]
I'll make a PR against your fork in a few hours

**Comment by blyxyas** [comment]
Just before the PR, could you create a new branch named "path_join_correct" and change this PR to point to that new branch (copy-pasting these changes)? Making a PR from the master branch will have some problems.

**Comment by ofeeg** [comment]
> Just before my PR, could you create a new branch named "path_join_correct" and change this PR to point to that new branch (copy-pasting these changes)? Making a PR from the master branch will have some problems.

Do I have to make a new PR to change the source branch? Github doesn't seem to just let me change ofeeg:master to ofeeg:path_join_correct.

**Comment by blyxyas** [comment]
No, you dont need make a new PR. I'm on mobile currently, but I'm pretty sure it's possible as I've done it for a few PRs

**Comment by ofeeg** [comment]
> No, you dont need make a new PR. I'm on mobile currently, but I'm pretty sure it's possible as I've done it for a few PRs


All the sources I'm reading are telling me I have to make a new PR and close this one in order to merge from a different branch.  If you know how to do it let me know; otherwise I will make a new PR and reference this one. I'll remember to not try to merge from my forks master in the future to prevent this.

**Comment by xFrednet** [comment]
As far as I know, there is no way to change the base branch of a PR. You can only force push another branch to the source branch of a PR, that would override the changes on the branch. But this PR is linked to `ofeeg:master`. I'm sorry. You can create a new PR or just continue development on your master branch, both options are totally file for me. :) @blyxyas Do you have a preference in this regard?

**Comment by blyxyas** [comment]
I have a preference for creating a new branch. Having clean git history is kinda very important when working in Rust (my first Pr had 110 commits, 75 comments of me fighting with Clippy, deleting the whole fork and remaking everything 😢)

From personal experience, it isn't something fun :(

Yeah, I'm not currently with a pc at range, but although I'm pretty sure it's possible, creating a new PR is also fine.

Make sure to use `r? xFrednet` in the PR description :)

**Pull request closed without merging**